### PR TITLE
feat: correctly dismantle ingester gRPC reponse in querier

### PR DESCRIPTION
**Pull request by crepererum**
This now correctly processes record batches for the different
partitions. The actual code change is rather small but it requires some
substantial test infrastructure.

~On top of #4320.~

### CODE REVIEW

**Comment by alamb** [comment]
I rebased this PR now that https://github.com/influxdata/influxdb_iox/pull/4320 is merged

File: querier/src/ingester/flight_client.rs [hunk 1]
```diff
@@ -88,14 +93,64 @@ impl FlightClient for FlightClientImpl {
...
         let request: flight::generated_types::IngesterQueryRequest =
             request.try_into().context(CreatingRequestSnafu)?;
 
-        client.perform_query(request).await.context(FlightSnafu)
+        let perform_query = client.perform_query(request).await.context(FlightSnafu)?;
+        Ok(Box::new(perform_query))
+    }
+}
+
+/// Data that is returned by an ingester gRPC query.
```
Comment (alamb) [inline querier/src/ingester/flight_client.rs:109]: 👍

File: querier/src/ingester/mod.rs [hunk 2]
```diff
@@ -48,6 +48,25 @@ pub enum Error {
         ingester_address: String,
         source: FlightClientError,
     },
+
+    #[snafu(display("Batch for misses partition id '{}'", ingester_address))]
```
Comment (alamb) [inline querier/src/ingester/mod.rs:52]:
```suggestion
    #[snafu(display("Batch is missing partition id from '{}'", ingester_address))]
```

File: querier/src/ingester/mod.rs [hunk 3]
```diff
@@ -48,6 +48,25 @@ pub enum Error {
...
+    MissingPartitionId { ingester_address: String },
+
+    #[snafu(display("Batch has malformed partition id '{}': {}", ingester_address, source))]
```
Comment (alamb) [inline querier/src/ingester/mod.rs:55]:
```suggestion
    #[snafu(display("Batch has malformed partition id from '{}': {}", ingester_address, source))]
```

Comment (alamb) [reply querier/src/ingester/mod.rs:55]: I just suggest making this consistent with the wording of `UnknownPartitionId` -- very minor

File: querier/src/ingester/mod.rs [hunk 4]
```diff
@@ -147,30 +166,62 @@ impl IngesterConnection for IngesterConnectionImpl {
...
-
-            let ingester_partition = IngesterPartition::try_new(
-                ChunkId::new(),
-                Arc::clone(&namespace_name),
-                Arc::clone(&table_name),
-                Arc::clone(&expected_schema),
-                parquet_max_sequence_number,
-                batches,
-            )?;
+            // read unpersisted partitions
```
Comment (alamb) [inline querier/src/ingester/mod.rs:169]:
```suggestion
            // read unpersisted partitions
            // map partition_id -> (PartitionMetadata, Vec<PartitionData>))
```

File: querier/src/ingester/mod.rs [hunk 5]
```diff
@@ -147,30 +166,62 @@ impl IngesterConnection for IngesterConnectionImpl {
...
+                    })?
+                    .1
+                    .push(batch);
+                num_batches += 1;
+            }
+            debug!(num_batches, "Received batches from ingester");
+            trace!(?partitions, schema=?perform_query.schema(), "Detailed from ingester");
+
+            for (partition_id, (state, batches)) in partitions {
+                // do NOT filter out empty partitions, because the caller of this functions needs the attached metadata
```
Comment (alamb) [inline querier/src/ingester/mod.rs:207]: 👍

File: querier/src/ingester/mod.rs [hunk 6]
```diff
@@ -432,23 +489,317 @@ mod tests {
...
+                },
+                schema: schema().as_arrow(),
+            }),
+        )]));
+        let ingester_conn = mock_flight_client.ingester_conn().await;
+        let err = get_partitions(&ingester_conn).await.unwrap_err();
+        assert_matches!(err, Error::RemoteQuery { .. });
+    }
+
+    #[tokio::test]
```
Comment (alamb) [inline querier/src/ingester/mod.rs:511]: I love the testing A+ ✅

File: querier/src/ingester/mod.rs [hunk 7]
```diff
@@ -432,23 +489,317 @@ mod tests {
...
+
+    fn record_batch(lp: &str, partition_id: i64) -> RecordBatch {
+        batch_md(
+            &lp_to_record_batch(lp),
+            HashMap::from([(String::from("iox:partition_id"), partition_id.to_string())]),
+        )
+    }
+
+    fn lp_to_record_batch(lp: &str) -> RecordBatch {
+        lp_to_mutable_batch(lp).1.to_arrow(Selection::All).unwrap()
```
Comment (alamb) [inline querier/src/ingester/mod.rs:763]: this is a clever way to make a record batch. Nice

**Review by alamb — APPROVED** [review]
Very nice

**Comment by crepererum** [comment]
Integration are failing for a good reason, see #4347.

**Pull request merged**